### ISSUE TITLE
Bug: Documentation comments are reported as single line comments in command line compilation

### ISSUE THREAD

**Issue by ghost** [dotnet/roslyn#10003]
We have an analyzer that goes through all single-line comments. In some cases this analyzer  reports on XML comments as well inside Visual Studio. The issue only happens if the analyzer is used as a NuGet package, and only after restoring the package with a build.

I have created a minimal analyzer that produces the same issue. The analyzer can be cloned from here https://github.com/tvsonar/doc-comment-repro

**Steps to reproduce**
1. Build the analyzer, 
2. Create a project with an XML comment in it and add the analyzer as a NuGet package
3. Close VS
4. Delete the `packages` folder from the disk
5. Open the solution in VS
6. Hit Shift+Ctrl+B, this will restore the package before the build
7. An issue shows up for each line of the XML comment
8. Close VS
9. Open the solution in VS
10. At this time only single line comments are displayed

This is how it looks in VS:
![actual](https://cloud.githubusercontent.com/assets/11485373/13982662/95758344-f0eb-11e5-9c14-a880de4704d3.png)

**Expected Behavior**:
Only single line comments should be reported in step 7.

This is after a VS restart:
![expected](https://cloud.githubusercontent.com/assets/11485373/13982667/9fc80f56-f0eb-11e5-88b0-84f6dad66d16.png)

The analyzer registers a `SyntaxTreeAction`, and traverses all the descendant tokens. For each token it checks the leading and trailing trivias, and report if it's a `SingleLineCommentTrivia`. XML comments should not be reported as their kind is `SingleLineDocumentationCommentTrivia`.

**Update**
This issues seem to be always reported if the build is done in the command line, and not inside VS.

**Comment by mavasani** [comment]
So the issue here is specific to command line compilation and default DocumentationMode. Command line compiler uses [DocumentationMode.None](http://source.roslyn.io/#Microsoft.CodeAnalysis/DocumentationMode.cs,18) as the default documentation mode, and hence ends up parsing XML doc comments in code as single line comments. Basically [this](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/Lexer.cs,2221) checks fails and its ends up treating doc comments as regular comments. This causes analyzer to receive 4 single line comments rather than 1 in the above file, and it reports false diagnostics. This issue doesn't show up in the IDE as it always uses [DocmentationMode.Diagnose](http://source.roslyn.io/#Microsoft.CodeAnalysis/DocumentationMode.cs,4dca93962a488709,references) to parse and report doc comment diagnostics. Similarly, turning on XML documentation option for this project causes these false analyzer diagnostics to not show up.

I think the compiler should use `DocumentationMode.Parse` as the default documentation mode, but probably compiler team knows why it is not so currently.
Tagging @jaredpar @gafter as this is a compiler bug. Also tagging @srivatsn as we have discussed this few times before but I can't recall the full context for punting this.

**Comment by mavasani** [comment]
Default documentation mode for command line compilers is set [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCommandLineParser.cs,1136)(C#) and [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/CommandLine/VisualBasicCommandLineParser.vb,1237)(VB)

**Comment by gafter** [comment]
I would be concerned about compiler throughput - taking a performance hit parsing doc comments. It is worth studying.

**Comment by Rantanen** [comment]
I encountered this same issue as well.

I'm fine with the workaround of requiring the XML documentation option set on the project level - however I would like to suppress the warnings in case this isn't set and instead raise a single "Enable XML documentation generation in the project" warning.

Is there a way to read the DocumentationMode value during analysis? I couldn't find a way through the AnalysisContext supplied to the DiagnosticAnalyzers.

**Comment by Rantanen** [comment]
Nevermind! Just found it under `SyntaxNodeAnalysisContext.Node.SyntaxTree.Options` >.>

**Comment by sharwell** [comment]
Superseded by #26938
